app/src/main/java/net/vonforst/evmap/storage/CacheLiveData.kt

```diff
@@ -0,0 +1,62 @@
+package net.vonforst.evmap.storage
+
+import android.util.Log
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MediatorLiveData
+import net.vonforst.evmap.viewmodel.Resource
+import net.vonforst.evmap.viewmodel.Status
+
+/**
+ * LiveData implementation that allows loading data both from a cache and an API.
+ *
+ * It gives the cache result while loading, and then switches to the API result if the API call was
+ * successful.
+ */
+class CacheLiveData<T>(cache: LiveData<T>, api: LiveData<Resource<T>>) :
+    MediatorLiveData<Resource<T>>() {
+    private var cacheResult: T? = null
+    private var apiResult: Resource<T>? = null
+
+    init {
+        updateValue()
+        addSource(cache) {
+            cacheResult = it
+            updateValue()
+        }
+        addSource(api) {
+            apiResult = it
+            updateValue()
+        }
+    }
+
+    private fun updateValue() {
+        val api = apiResult
+        val cache = cacheResult
+
+        if (api == null && cache == null) {
```

Comment:
you could create a method for each case for example: 
`
fun isStillLoading(): Boolean = api == null && cache == null
`
that could improve code readability.